### PR TITLE
Fix/password requirements

### PR DESCRIPTION
**Pull request by marcelluscaio**
- Adapted Check component to accept no highlighted text, avoiding code duplication
- Changed joi validation for password on Client and on Server, so they are validating on the same rules
- Deleted all validation done without the use of joi
- Created custom hook to keep validation of password and confirm in the same place
- Applied new architecture to SetNewPassword
- Broke Register Steps in different components (Step one two and three)
- Applied new architecture on Step Three
-  Applied new architecture on Login
- Applied new architecture on  Password Panel
- Adjusted styles on all of the pages 
- Password Tests
  - Valid Password and Matching Confirmation
  - Valid Password, Incorrect Confirmation, then Corrects confirmation
  - Password Update to Match Confirm Password
  - Short Password with Matching Confirmation
  - Invalid Password Format, Correct Confirmation
  - Empty Password Field with Confirmation Filled
  - Confirm Password Left Blank
  - Password Change After Successful Match
- Implemented keyboard navigation and semantic forms on the form I touchedm addressing issue #1105 partially
- Got rid of negative tabindexes on interactive items

Closes Issue #1021

### CODE REVIEW

File: Client/src/Pages/Auth/Register/StepThree/index.jsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+						<Check
+							noHighlightText={"Must contain at least"}
+							text={"one lower character"}
+							variant={feedbacks.lowercase}
+						/>
+						<Check
+							noHighlightText={"Confirm password and password"}
+							text={"must match"}
+							variant={feedbacks.confirm}
+						/>
```
Comment (ajhollid) [inline Client/src/Pages/Auth/Register/StepThree/index.jsx:116]: This looks really great, I love the improvement here :+1:

File: Client/src/Pages/Auth/Register/StepThree/index.jsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+							Back
+						</Button>
+						<Button
+							type="submit"
+							variant="contained"
+							color="primary"
+							disabled={
+								form.password.length === 0 ||
+								form.confirm.length === 0 ||
+								Object.keys(errors).length !== 0
```
Comment (ajhollid) [inline Client/src/Pages/Auth/Register/StepThree/index.jsx:148]: Appreciate the strict evalutations here, not a big fan of JavaScripts type coercion

File: Client/src/Pages/Auth/Register/StepOne/index.jsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+						variant="outlined"
+						color="info"
+						onClick={onBack}
+						sx={{
+							px: theme.spacing(5),
+							"& svg.MuiSvgIcon-root": {
+								mr: theme.spacing(3),
+							},
+						}}
+						props={{ tabIndex: -1 }}
```
[the remark this answers is not in the document]
Comment (ajhollid) [reply Client/src/Pages/Auth/Register/StepOne/index.jsx:101]: As far as I can tell from [the docs](https://mui.com/material-ui/api/button/#props) the `Button` component doesn't expose `props` prop, nor does the `BaseButton` API :thinking:  It does expose `tabIndex` though, was there a reason not to use the `tabIndex` prop?

File: Server/validation/joi.js [hunk 4]
```diff
@@ -18,6 +18,9 @@ const roleValidatior = (role) => (value, helpers) => {
 // Auth
 //****************************************
 
+const passwordPattern =
+	/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!?@#$%^&*()\-_=+[\]{};:'",.<>~`|\\/])[A-Za-z0-9!?@#$%^&*()\-_=+[\]{};:'",.<>~`|\\/]+$/;
+
```
Comment (ajhollid) [inline Server/validation/joi.js:23]: Apprecaite the refactor here!

**Review by ajhollid — CHANGES_REQUESTED** [review]
Overall it looks great, I love the new feedback for users when setting up a password, much more informative and user friendly.

There's an issue with the Toasts that needs to be resovled, and I'm not sure of the reasoning for using the `props` prop for the `Button` components as it doesn't appear to expose that prop.

If I've missed something with regards to the `Button` let me know!

Other than that looks good to me!

File: Client/src/Validation/validation.js [hunk 5]
```diff
@@ -61,12 +68,12 @@ const credentials = joi.object({
 		.string()
 		.trim()
 		.messages({
-			"string.empty": "Password confirmation is required",
+			"string.empty": "empty",
```
[the remark this answers is not in the document]
Comment (marcelluscaio) [reply Client/src/Validation/validation.js:71]: Great point. The idea was not allowing submission while there were errors. But I also felt weird while using messages to cast error types. I'll look into that, and won't mix error type-casting and messages

File: Client/src/Pages/Auth/Register/StepOne/index.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcelluscaio) [reply Client/src/Pages/Auth/Register/StepOne/index.jsx:101]: Solved!

**Comment by marcelluscaio** [comment]
@ajhollid I think I addressed all of the problems. Most important, I was able to detach  joi's error type from message. Thank you again for your thorough review =D

**Comment by ajhollid** [comment]
> @ajhollid I think I addressed all of the problems. Most important, I was able to detach  joi's error type from message. Thank you again for your thorough review =D

Awesome, nice work! I will review as soon as I can 🙏

**Comment by marcelluscaio** [comment]
I updated the description by the way, Any doubt, let me know =)

File: Client/src/Validation/validation.js [hunk 6]
```diff
@@ -16,25 +16,37 @@ const passwordSchema = joi
...
-
 		return value;
+	})
+	.messages({
+		"string.empty": "Password is required",
+		"string.min": "Password must be at least 8 characters long",
+		uppercase: "Password must contain at least one uppercase letter",
+		lowercase: "Password must contain at least one lowercase letter",
+		number: "Password must contain at least one number",
+		special: "Password must contain at least one special character",
```
Comment (ajhollid) [inline Client/src/Validation/validation.js:49]: Looking good :+1:  Nice work digging through the docs to figure this one out

**Review by ajhollid — APPROVED** [review]
Great work, everything looks good to me and is working as expected.

I really like being able to login/register without having to touch the mouse now :+1:

**Pull request merged**